### PR TITLE
Use baro singleton

### PR DESCRIPTION
**Pull request by peterbarker**
Some time ago we created a barometer singleton.

This PR makes us use it everywhere except within Baro itself.  Notionally we could do without the `_frontend` AP_Baro_Backend member, but it looks odd doing that.

### CODE REVIEW

File: APMrover2/AP_Arming.h [hunk 1]
```diff
@@ -9,16 +9,16 @@
...
                     const AP_BattMonitor &battery, const AC_Fence &fence)
-        : AP_Arming(ahrs_ref, baro, compass, battery),
+        : AP_Arming(ahrs_ref, compass, battery),
           _fence(fence)
     {
     }
 
     /* Do not allow copies */
     AP_Arming_Rover(const AP_Arming_Rover &other) = delete;
-    AP_Arming_Rover &operator=(const AP_Baro&) = delete;
```
Comment (khancyr) [inline APMrover2/AP_Arming.h]: that doesn't look right  and nobody saw it ...

File: libraries/AP_NavEKF2/AP_NavEKF2_Measurements.cpp [hunk 2]
```diff
@@ -556,10 +556,11 @@ void NavEKF2_core::readBaroData()
 {
     // check to see if baro measurement has changed so we know if a new measurement has arrived
     // do not accept data at a faster rate than 14Hz to avoid overflowing the FIFO buffer
-    if (frontend->_baro.get_last_update() - lastBaroReceived_ms > 70) {
+    AP_Baro &baro = AP::baro();
```
Comment (OXINARF) [inline libraries/AP_NavEKF2/AP_NavEKF2_Measurements.cpp]: const?

File: libraries/AP_NavEKF3/AP_NavEKF3_Measurements.cpp [hunk 3]
```diff
@@ -603,10 +603,10 @@ void NavEKF3_core::readBaroData()
 {
     // check to see if baro measurement has changed so we know if a new measurement has arrived
     // limit update rate to avoid overflowing the FIFO buffer
-    if (frontend->_baro.get_last_update() - lastBaroReceived_ms > frontend->sensorIntervalMin_ms) {
+    if (AP::baro().get_last_update() - lastBaroReceived_ms > frontend->sensorIntervalMin_ms) {
```
Comment (OXINARF) [inline libraries/AP_NavEKF3/AP_NavEKF3_Measurements.cpp]: I'd prefer we keep the same style in EKF2 and EKF3, either both get a variable or both don't.

File: libraries/DataFlash/LogFile.cpp [hunk 4]
```diff
@@ -425,8 +425,9 @@ void DataFlash_Class::Log_Write_RSSI(AP_RSSI &rssi)
     WriteBlock(&pkt, sizeof(pkt));
 }
 
-void DataFlash_Class::Log_Write_Baro_instance(AP_Baro &baro, uint64_t time_us, uint8_t baro_instance, enum LogMessages type)
+void DataFlash_Class::Log_Write_Baro_instance(uint64_t time_us, uint8_t baro_instance, enum LogMessages type)
 {
+    AP_Baro &baro = AP::baro();
```
Comment (OXINARF) [inline libraries/DataFlash/LogFile.cpp]: const?

File: libraries/DataFlash/LogFile.cpp [hunk 5]
```diff
@@ -445,17 +446,18 @@ void DataFlash_Class::Log_Write_Baro_instance(AP_Baro &baro, uint64_t time_us, u
...
 
 // Write a BARO packet
-void DataFlash_Class::Log_Write_Baro(AP_Baro &baro, uint64_t time_us)
+void DataFlash_Class::Log_Write_Baro(uint64_t time_us)
 {
     if (time_us == 0) {
         time_us = AP_HAL::micros64();
     }
-    Log_Write_Baro_instance(baro, time_us, 0, LOG_BARO_MSG);
+    AP_Baro &baro = AP::baro();
```
Comment (OXINARF) [inline libraries/DataFlash/LogFile.cpp]: const?

File: libraries/GCS_MAVLink/GCS_Common.cpp [hunk 6]
```diff
@@ -1062,9 +1062,10 @@ void GCS_MAVLINK::send_raw_imu(const AP_InertialSensor &ins, const Compass &comp
         mag.z);        
 }
 
-void GCS_MAVLINK::send_scaled_pressure(AP_Baro &barometer)
+void GCS_MAVLINK::send_scaled_pressure()
 {
     uint32_t now = AP_HAL::millis();
+    AP_Baro &barometer = AP::baro();
```
Comment (OXINARF) [inline libraries/GCS_MAVLink/GCS_Common.cpp]: const?

File: libraries/GCS_MAVLink/GCS_Common.cpp [hunk 7]
```diff
@@ -1110,6 +1111,8 @@ void GCS_MAVLINK::send_sensor_offsets(const AP_InertialSensor &ins, const Compas
     const Vector3f &accel_offsets = ins.get_accel_offsets(0);
     const Vector3f &gyro_offsets = ins.get_gyro_offsets(0);
 
+    AP_Baro &barometer = AP::baro();
```
Comment (OXINARF) [inline libraries/GCS_MAVLink/GCS_Common.cpp]: const?

File: ArduCopter/GCS_Mavlink.cpp [hunk 8]
```diff
@@ -1662,7 +1662,7 @@ void GCS_MAVLINK_Copter::handleMessage(mavlink_message_t* msg)
 
         ins.set_accel(0, accels);
 
-        copter.barometer.setHIL(packet.alt*0.001f);
+        AP::baro().barometer.setHIL(packet.alt*0.001f);
```
Comment (OXINARF) [inline ArduCopter/GCS_Mavlink.cpp]: Should just be `AP::baro().setHil` right?

File: Tools/Replay/LR_MsgHandler.h [hunk 9]
```diff
@@ -113,13 +113,13 @@ class LR_MsgHandler_BARO : public LR_MsgHandler
...
     LR_MsgHandler_BARO(log_Format &_f, DataFlash_Class &_dataflash,
-                    uint64_t &_last_timestamp_usec, AP_Baro &_baro)
-        : LR_MsgHandler(_f, _dataflash, _last_timestamp_usec), baro(_baro) { };
+                    uint64_t &_last_timestamp_usec)
+        : LR_MsgHandler(_f, _dataflash, _last_timestamp_usec)
+        { };
 
     virtual void process_message(uint8_t *msg);
 
 private:
```
Comment (OXINARF) [inline Tools/Replay/LR_MsgHandler.h]: Remove this?

**Comment by peterbarker** [comment]
On Wed, 7 Mar 2018, Francisco Ferreira wrote:

> I've marked quite a few places for const, but I'm guessing that's not
> possible because methods aren't marked as const? I didn't check if they
> could be marked const, if they could it would be great.

Sadly, some of the getters update internal filters.

Which means that the filter rate changes based on your mavlink stream 
rates.

This is not a feature.

File: libraries/AP_Baro/AP_Baro.cpp [hunk 10]
```diff
@@ -679,3 +679,11 @@ void AP_Baro::set_pressure_correction(uint8_t instance, float p_correction)
     }
 }
 
+namespace AP {
+
+AP_Baro &baro()
+{
+    return *AP_Baro::get_instance();
```
Comment (lucasdemarchi) [inline libraries/AP_Baro/AP_Baro.cpp]: any reason why this kind of getter can't be moved to the header?

File: libraries/DataFlash/LogFile.cpp (hunk 4 above)
Comment (lucasdemarchi) [reply libraries/DataFlash/LogFile.cpp]: it calls `baro.get_climb_rate()` that uses DerivativeFilter::slope() which modifies its state. So we can't propagate constness, at least not in this PR.

**Comment by lucasdemarchi** [comment]
\o/

**Comment by peterbarker** [comment]
On Fri, 9 Mar 2018, Lucas De Marchi wrote:

> +namespace AP {
> +
> +AP_Baro &baro()
> +{
> +    return *AP_Baro::get_instance();
> 
> any reason why this kind of getter can't be moved to the header?

As in tacked onto the bottom of the header?  Won't you end up with 
redefinition errors in that case?

Yep:

AP_MotorsTri.cpp:(.text._ZN2AP4baroEv+0x0): multiple definition of 
`AP::baro()'
ArduCopter/AP_Arming.cpp.20.o:AP_Arming.cpp:(.text._ZN2AP4baroEv+0x0): 
first defined here


> >  {
> +    AP_Baro &baro = AP::baro();
> 
> it calls baro.get_climb_rate() that uses DerivativeFilter::slope() which
> modifies its state. So we can't propagate constness, at least not in this
> PR.

Always something to look forward to :-)

**Pull request closed without merging**